### PR TITLE
Review code to prepare ANSSI submission

### PR DESCRIPTION
**Pull request by tbrezot**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,18 @@
...
+---
+## [1.0.0] - 2022-07-27
+### Added
+### Changed
+- Add doc
+- Several object internals have been made private
+- review code for ANSSI submission
+### Fixed
+### Removed
+- `ap()` is replaced by `AccessPolicy::new()`
```
Comment (ThibsG) [inline CHANGELOG.md]: can you update test examples in doc headers please, they don't compile anymore (`cargo test --doc`)

File: src/attribute.rs [hunk 2]
```diff
@@ -1,13 +1,11 @@
 #![allow(clippy::module_name_repetitions)]
```
Comment (ThibsG) [inline src/attribute.rs]: Not necessary
```suggestion
```

File: src/access_policy.rs [hunk 3]
```diff
@@ -1,21 +1,20 @@
 #![allow(clippy::module_name_repetitions)]
```
Comment (ThibsG) [inline src/access_policy.rs]: Not necessary
```suggestion
```

File: src/lib.rs [hunk 4]
```diff
@@ -1,11 +1,20 @@
+//! This crate defines the `Policy` logic, the basis for Attribute Based
+//! Encryption (ABE).
+//!
+//! A `Policy` is a set of axes. Each axis is defined by its name and its list
+//! of associated attribute names.
+//!
+//! An `Attribute` is composed by an axis name and an attribute name within
+//! this axis.
+
 #![allow(clippy::module_name_repetitions)]
```
Comment (ThibsG) [inline src/lib.rs]: Not necessary
```suggestion
```

File: src/policy.rs [hunk 5]
```diff
@@ -1,18 +1,16 @@
 #![allow(clippy::module_name_repetitions)]
```
Comment (ThibsG) [inline src/policy.rs]: Not necessary
```suggestion
```

File: src/policy.rs [hunk 6]
```diff
@@ -30,36 +34,42 @@ impl PolicyAxis {
         }
     }
 
+    /// Returns the number of attributes belonging to this axis.
     #[allow(clippy::len_without_is_empty)]
```
Comment (ThibsG) [inline src/policy.rs]: I think it is not so much effort to add an `is_empty` method as recommended by Clippy. What do you think?

File: src/policy.rs [hunk 7]
```diff
@@ -21,6 +19,12 @@ pub struct PolicyAxis {
 }
 
 impl PolicyAxis {
+    /// Generate a new policy axis with the given name and attribute names. A
+    /// hierarchical axis enforce order between its attributes.
```
Comment (ThibsG) [inline src/policy.rs]:
```suggestion
    /// Generates a new policy axis with the given name and attribute names.
    /// A hierarchical axis enforces order between its attributes.
```

File: src/policy.rs [hunk 8]
```diff
@@ -30,36 +34,42 @@ impl PolicyAxis {
...
     pub fn attributes(&self) -> &[String] {
         &self.attributes
     }
 
+    /// Returns the axis name.
     pub fn name(&self) -> &str {
         &self.name
     }
 
+    /// Return `true` if this axis is hierarchical.
```
Comment (ThibsG) [inline src/policy.rs]:
```suggestion
    /// Returns `true` if this axis is hierarchical.
```

File: src/policy.rs [hunk 9]
```diff
@@ -30,36 +34,42 @@ impl PolicyAxis {
...
-    pub max_attribute_value: u32,
-    /// store the policies by name
-    pub store: HashMap<String, (Vec<String>, bool)>,
+    /// Last value taken by the attriute.
+    last_attribute_value: u32,
+    /// Maximum attribute value. Defines a maximum number of attribute
+    /// creations (revocations + addition).
+    max_attribute_value: u32,
+    /// Policy axes
+    axes: HashMap<String, (Vec<String>, bool)>,
```
Comment (ThibsG) [inline src/policy.rs]: can we add details about the content, especially on the right side. Or another idea, create a type alias for this one

File: src/policy.rs [hunk 10]
```diff
@@ -73,35 +83,34 @@ impl Display for Policy {
...
-    /// Returns the policy in the for of a Map where
+    /// Returns the policy in the form of a Map where
     ///  - the keys are the axis names
     ///  - the values are a tuple of
     ///     - list of attribute names for that axis
     ///     - whether the axis hierarchical
     pub fn as_map(&self) -> &HashMap<String, (Vec<String>, bool)> {
-        &self.store
+        &self.axes
     }
```
Comment (ThibsG) [inline src/policy.rs]: doesn't seem useful, make back the item `pub`

File: src/policy.rs [hunk 11]
```diff
@@ -30,36 +34,42 @@ impl PolicyAxis {
...
 pub struct Policy {
-    pub last_attribute_value: u32,
-    pub max_attribute_value: u32,
-    /// store the policies by name
-    pub store: HashMap<String, (Vec<String>, bool)>,
+    /// Last value taken by the attriute.
+    last_attribute_value: u32,
+    /// Maximum attribute value. Defines a maximum number of attribute
+    /// creations (revocations + addition).
+    max_attribute_value: u32,
```
Comment (ThibsG) [inline src/policy.rs]: shouldn't this be `max_attribute_creations`?

File: src/policy.rs [hunk 12]
```diff
@@ -73,35 +83,34 @@ impl Display for Policy {
...
     pub fn as_map(&self) -> &HashMap<String, (Vec<String>, bool)> {
-        &self.store
+        &self.axes
     }
 
+    /// Return the number of attribute creations allowed.
     #[must_use]
     pub fn max_attr(&self) -> u32 {
         self.max_attribute_value
     }
```
Comment (ThibsG) [inline src/policy.rs]: if we rename `max_attribute_value` as suggested, just make it `pub`.

Also, what do you think about adding a new method: `pub fn remaining_attribute_creations(&self)`?

File: src/policy.rs [hunk 13]
```diff
@@ -136,8 +145,8 @@ impl Policy {
         Ok(())
     }
 
-    /// Rotate an attribute, changing its underlying value with that of an
-    /// unused slot
+    /// Rotates an attribute, changing its underlying value with that of an
+    /// unused slot.
```
Comment (ThibsG) [inline src/policy.rs]: This comment is confusing, I don't understand the: `with that of an unused slot`.

`marking its underlying value as deprecated` is what it means?

File: src/policy.rs (hunk 6 above)
Comment (tbrezot) [reply src/policy.rs]: Strange... I have no clippy warnings

File: src/policy.rs (hunk 9 above)
Comment (tbrezot) [reply src/policy.rs]: I change the  name to make it more clear: these are axes, so axis name, attribute names and `is_hierarchical`. But I can document it.

File: src/policy.rs (hunk 10 above)
Comment (tbrezot) [reply src/policy.rs]: Okay. What are the good practices about attribute visibility ?

File: src/policy.rs (hunk 11 above)
Comment (tbrezot) [reply src/policy.rs]: It can, indeed.

File: src/policy.rs (hunk 12 above)
Comment (tbrezot) [reply src/policy.rs]: this can be done :)

File: src/policy.rs (hunk 13 above)
Comment (tbrezot) [reply src/policy.rs]: No, it means the new value will be a value not already taken by any attribute. But I agree, this is not clear, I will change.

File: src/access_policy.rs [hunk 14]
```diff
@@ -357,7 +360,9 @@ impl AccessPolicy {
         }
     }
 
-    /// Retrieve all the Attributes present in this access policy
+    /// Retrieves all the Attributes present in this access policy.
```
Comment (ThibsG) [inline src/access_policy.rs]:
```suggestion
    /// Retrieves all the attributes present in this access policy.
```

File: src/policy.rs [hunk 15]
```diff
@@ -30,36 +32,46 @@ impl PolicyAxis {
...
 
-    #[allow(clippy::len_without_is_empty)]
+    /// Returns the number of attributes belonging to this axis.
     #[must_use]
     pub fn len(&self) -> usize {
         self.attributes.len()
     }
 
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
```
Comment (ThibsG) [inline src/policy.rs]:
```suggestion
        self.attributes.is_empty()
```

File: src/policy.rs [hunk 16]
```diff
@@ -30,36 +32,46 @@ impl PolicyAxis {
...
-    pub max_attribute_value: u32,
-    /// store the policies by name
-    pub store: HashMap<String, (Vec<String>, bool)>,
+    /// Last value taken by the attriute.
+    last_attribute_value: u32,
+    /// Maximum attribute value. Defines a maximum number of attribute
+    /// creations (revocations + addition).
+    pub max_attribute_creations: u32,
+    /// Policy axes: maps axes name to the list of associated attribute names
+    /// and a bollean defining whether or not this axis is hierarchical.
```
Comment (ThibsG) [inline src/policy.rs]:
```suggestion
    /// and a boolean defining whether or not this axis is hierarchical.
```

File: src/policy.rs [hunk 17]
```diff
@@ -30,36 +32,46 @@ impl PolicyAxis {
...
 
+    /// Returns the axis name.
     pub fn name(&self) -> &str {
         &self.name
     }
 
+    /// Returns `true` if this axis is hierarchical.
     pub fn is_hierarchical(&self) -> bool {
         self.hierarchical
     }
```
Comment (ThibsG) [inline src/policy.rs]: I don't think these are useful. Just make items public

File: src/policy.rs [hunk 18]
```diff
@@ -73,47 +85,40 @@ impl Display for Policy {
...
-            max_attribute_value: nb_revocation,
-            store: HashMap::new(),
+            max_attribute_creations: nb_creations,
+            axes: HashMap::new(),
             attribute_to_int: HashMap::new(),
         }
     }
 
-    /// Returns the policy in the for of a Map where
+    pub fn remaining_attribute_creations(&self) -> u32 {
```
Comment (ThibsG) [inline src/policy.rs:90]: missing doc header

File: CHANGELOG.md (hunk 1 above)
Comment (ThibsG) [inline CHANGELOG.md]:
```suggestion
- `ap()` is removed. Use `AccessPolicy::new()` instead
```

File: src/policy.rs (hunk 10 above)
Comment (ThibsG) [reply src/policy.rs]: don't write getters that just returns a reference on items without modifying them.

User has already the right to manipulate the struct, so it is straight forward to make the item `pub` and require less code.

File: src/policy.rs [hunk 19]
```diff
@@ -30,36 +32,46 @@ impl PolicyAxis {
...
 pub struct Policy {
-    pub last_attribute_value: u32,
-    pub max_attribute_value: u32,
-    /// store the policies by name
-    pub store: HashMap<String, (Vec<String>, bool)>,
+    /// Last value taken by the attriute.
+    last_attribute_value: u32,
+    /// Maximum attribute value. Defines a maximum number of attribute
+    /// creations (revocations + addition).
+    pub max_attribute_creations: u32,
```
Comment (Manuthor) [inline src/policy.rs:58]: it is a breaking change. Renaming this impacts the serialization: the TTLV serialization in KMS must be updated, and so the cosmian_java_lib. Also it impacts the WASM and hence the cosmian_js_lib. 
Is it what we want?

File: CHANGELOG.md (hunk 1 above)
Comment (Manuthor) [reply CHANGELOG.md]: I am pretty sure than `ap` function has been made for practical usage (and to avoid to write `AccessPolicy::new()`)

**Review by Manuthor — COMMENTED** [review]
Renaming the Policy attributes comes with a price (impacting other libs)

File: src/policy.rs (hunk 17 above)
Comment (tbrezot) [reply src/policy.rs]: ok

File: src/policy.rs (hunk 19 above)
Comment (ThibsG) [reply src/policy.rs:58]: the version will reflect the breaking change. Upgrading is not required absolutely.

File: CHANGELOG.md (hunk 1 above)
Comment (tbrezot) [reply CHANGELOG.md]: It is possible, but this means having two times the same code, and I don't think this is a good thing. I don't see why calling `AccessPolicy::new()` is not "practical".

File: src/policy.rs (hunk 18 above)
Comment (tbrezot) [reply src/policy.rs:90]: done

File: src/policy.rs (hunk 10 above)
Comment (tbrezot) [reply src/policy.rs]: ok

**Pull request merged**